performant-software/dm-2
# Email fix (breaking change)

### What this PR does

- Per #324:
  - Sends emails to all admins when a new user registers
- After registering, forces users to confirm their emails before logging in

### Additional notes

I originally thought the goal was email _confirmations_, i.e. a new user should get an email upon registration, which they must open to complete registration. That might be a desired feature, but can be disabled if not.

#### Breaking changes

- In order to not encounter SendGrid errors, it is required that system emails come from a valid email address. Thus, rather than determine this programmatically from the hostname (which can break as in the case of `localhost`-like hostnames), it has been set as an environment variable `EMAIL_FROM`. That environment variable must be set for emails to send, so it is considered a breaking change.
- Because users must confirm their email when registering, all users must now be marked "confirmed" in the DB, or else they cannot log in. Of course, all previous email addresses were unconfirmed, so this has to be done retroactively for existing accounts. I have created a migration helper that must be run on all instances to fix this problem:

```ruby
UserConfirmationMigrationHelper.confirm_approved_users!
```

If we decide not to go with the "email confirmation" feature then this migration is not necessary.

### Status


### Steps to test

1. Visit https://dm-2-staging.herokuapp.com/
2. Register for a new account with a valid email address

Issue #324:

3. Check with me that I got an email about your registration

Confirmation stuff:

4. Try to log in without confirming your email first and verify that an error appears
5. Then, confirm your email and verify that a popup appears
6. Log in with the new account
